### PR TITLE
 Add function to query windows known folders

### PR DESCRIPTION
**Pull request by musm**
As suggested and remarked by `@stevengj` relying on environment variables for system paths can be dangerous, especially since users can mess with them. 

For example, an easy way to break  julia's download function is as follows
```julia
ENV["SYSTEMROOT"] = "blah"

julia> download("https://github.com")                                                               
ERROR: IOError: could not spawn `'blah\System32\WindowsPowerShell\v1.0\powershell.exe'
```

In this PR we wrap windows api to access known folders from julia, to prevent such problems.

### CODE REVIEW

File: base/path.jl [hunk 1]
```diff
@@ -477,3 +477,73 @@ relpath(path::AbstractString, startpath::AbstractString) =
...
+"C:\\WINDOWS\\Fonts"
+```
+
+"""
+function get_known_folder_path(folderid::GUID)
+    pathptr = Ref{Ptr{Cwchar_t}}()
+    result = ccall((:SHGetKnownFolderPath,:shell32), stdcall, Cint,
+                    (GUID, UInt32, Ptr{Cvoid}, Ref{Ptr{Cwchar_t}}), folderid, KF_FLAG_DEFAULT, C_NULL, pathptr)
+    result != 0 && error()
+    pathbuf = unsafe_wrap(Vector{Cwchar_t}, pathptr[], ccall(:wcslen, UInt, (Ptr{Cwchar_t},), pathptr[]))
```
Comment (c42f) [inline base/path.jl]: `wcslen` returns `size_t`; best use `Csize_t` for clarity.

File: base/path.jl [hunk 2]
```diff
@@ -477,3 +477,73 @@ relpath(path::AbstractString, startpath::AbstractString) =
...
+    path = transcode(String, pathbuf)
+    ccall((:CoTaskMemFree, :ole32), stdcall, Cvoid, (Ptr{Cwchar_t},), pathptr[])
```
Comment (c42f) [inline base/path.jl]: Technically this takes `void*` as input.

**Review by c42f — COMMENTED** [review]
Agreed that environment variables aren't the most robust things to depend on. Was the motivation real or theoretical here? (Ie, did a user actually break their julia install by messing with `%SYSTEMROOT%`?)

Regarding API, this looks reasonable if it's limited to internal use by `Base`. If we wanted a general "standard paths" API, we should consider looking for some API inspiration elsewhere, for example:
* QStandardPath https://doc.qt.io/qt-5/qstandardpaths.html
* PlatformFolders https://github.com/sago007/PlatformFolders

This needs some kind of tests before merging. Obviously the output of this function will depend on the test machine configuration. That's annoying but you could do some pattern matching I suppose.

**Comment by musm** [comment]
@c42f thank you for the review and comments.

> Agreed that environment variables aren't the most robust things to depend on. Was the motivation real or theoretical here? (Ie, did a user actually break their julia install by messing with %SYSTEMROOT%?)

The concern here was more theoretical.

My initial motivation was just to limit this to internal use in Base, although I see the benefit of a better API for standardpaths.

**Comment by c42f** [comment]
Hum. If the concern is theoretical it does make me wonder how useful it is to have all this code to kill off a couple of environment variable uses which may be removed anyway (https://github.com/JuliaLang/julia/issues/27043). Do you have a link to the original conversation containing a bit more motivation for doing this?

**Comment by musm** [comment]
Well other packages that rely on these system paths, specifically PyCall,  directly call `SHGetKnownFolderPath` in order to avoid incorrect environment variables since it can be brittle. An internal API that can be used by such packages is desirable. 

These directories can be different from when run from   msys or cygwin terminal, and environment variables may be overridden.  SHGetKnownFolderPath() always gives the correct path to known system folders. 

There probably needs to be some more discussion on an API and how this should all look like, but I wanted to get the ball rolling with a PR.

**Comment by c42f** [comment]
Oh excellent: that's good motivation to have this code if it's being duplicated across the ecosystem. (I was just a little worried that it didn't seem entirely worthwhile to clean up a couple of uses in Base.)

However, I think that also means designing a public API which can be portable (to the extent possible) between systems. At a quick glance Qt's `QStandardPath` is a good place to look for design inspiration.

Another place to look for use cases (in addition to PyCall) would be the handling of standard paths rooted in the user home directories as used to find julia configuration, julia Pkg data, etc. How are they handled and can it be unified in a neat way under this API?

**Comment by musm** [comment]
Stale

**Comment by musm** [comment]
I put this in a package https://github.com/musm/WinKnownPaths.jl
where all the paths are available

**Comment by musm** [comment]
@c42f should we think about more general platform folder abstractions? I registered the julia package, but it might be beneficial to try to make a cross-platform version with the same idea.

**Comment by musm** [comment]
I'm just not sure if it's as a big an issue on macos and linux as windows ?

File: base/path.jl [hunk 3]
```diff
@@ -477,3 +477,73 @@ relpath(path::AbstractString, startpath::AbstractString) =
...
+"C:\\WINDOWS\\system32"
+
+julia> get_known_folder_path(FOLDERID.Fonts)
+"C:\\WINDOWS\\Fonts"
+```
+
+"""
+function get_known_folder_path(folderid::GUID)
+    pathptr = Ref{Ptr{Cwchar_t}}()
+    result = ccall((:SHGetKnownFolderPath,:shell32), stdcall, Cint,
```
Comment (musm) [inline base/path.jl:541]: For some reason I am unable to get this line to run on win-32bit.  It keeps spitting `ERROR: ReadOnlyMemoryError()`

**Pull request closed without merging**